reference/src/layout/arrays.md

```diff
@@ -0,0 +1,76 @@
+# Layout of Rust array types
+
+Array types, `[T; N]`, store `N` values of type `T` with a constant _stride_.
+Here, _stride_ is the distance between each pair of consecutive values within
+the array.
+
+The _offset_ of the first array element is `0`, that is, a pointer to the array
+and a pointer to its first element both point to the same memory address.
+
+The _alignment_ of array types is greater or equal to the alignment of its
+element type. If the element type is `repr(C)` the layout of the array is
+guaranteed to be the same as the layout of a C array with the same element type.
+
+> **Note**: the type of array arguments in C function signatures, e.g., `void
+> foo(T x[N])`, decays to a pointer. That is, these functions do not take arrays
+> as an arguments, they take a pointer to the first element of the array
+> instead. Array types are therefore _improper C types_ (not C FFI safe) in Rust
+> foreign function declarations, e.g., `extern { fn foo(x: [T; N]) -> [U; M];
+> }`. Pointers to arrays are fine: `extern { fn foo(x: *const [T; N]) -> *const
+> [U; M]; }`.
+
+The _stride_ of the array is the distance between a pair of consecutive values
+within the array, and it is constant for all element pairs. It is computed as
+the _size_ of the element type rounded up to the next multiple of the
+_alignment_ of the element type.
+
+When the element _size_ is a multiple of the element's _alignment_, then `stride
+== size`, and the elements are laid out contiguously in memory, e.g., `[u8; 4]`.
+In this case, the _size_ of the array can be computed as `size_of::<T>() * N`,
+and a pointer to the `i`-th element of the array can be obtained by offsetting a
+pointer to the first element of the array by `i`[^1].
+
+> **Note:** In the current Rust implementation, _size_ is always a multiple of
+> the element's _alignment_, and therefore `stride == size` always holds. This
+> is, however, not guaranteed by the [layout of structs and tuples].
+
+[^1]: When `stride > size` the pointer needs to be advanced by the array
+    _stride_ instead of by the element _size_.
+
+[layout of structs and tuples]: ./structs-and-tuples.md
+
+The [layout of Vector types][Vector] [^2] requires the _size_ and _alignment_ of
```

Comment:
I'd still prefer if we spoke about "SIMD vectors". But it seems I am alone with that position?

Reply:
No, I agree that the extra clarity is worth it here. The whole text should be consistent about this, though.

Reply:
I can rename the `Vector` layout to `Packed SIMD Vector` layout to make this clearer (and differentiate these from Cray Vectors).

Reply:
"cray vectors"? Now I am even more confused.^^ I also don't feel "packed" adds anything here because anyway there'd be no padding between the vector elements. So for an outsider it just seems redundant. Seems like you are trying to make a distinction with some obscure other kind of vectors that is long obsolete?

Remember you are not writing this text for people that are experts in vector types like you are. I wrote some manually vectorized code once and even I get confused by your terminology.

Reply:
I hope I clarified all uses of "vector" with "packed SIMD vector" in the last commit. There wasn't really a need to mention other vector types anywhere AFAICT. Better to worry about that when / if we add new vector types.